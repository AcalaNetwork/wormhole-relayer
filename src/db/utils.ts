import { isEvmAddress } from '@acala-network/eth-providers';
import { prisma } from './client';

export interface RouterInfo {
  params: string;       // stringified JSON params
  factoryAddr: string;
  recipient: string;
  feeAddr: string;
  routerAddr: string;
}

export const connectDb = async () => {
  try {
    await prisma.$connect();
  } catch (err) {
    console.error('❗️ db connection failed!');
    throw err;
  }
};

const _sanitizeParams = (params: string) => {
  let jsonParams: Record<string, string> = {};
  try {
    jsonParams = JSON.parse(params);
  } catch (err) {
    return params;
  }

  Object.keys(jsonParams).forEach(key => {
    if (isEvmAddress(jsonParams[key])) {
      jsonParams[key] = jsonParams[key].toLowerCase();
    }
  });

  return JSON.stringify(jsonParams);
};

// use lowercase for all addrs to be consistent
export const sanitizeRouterInfo = <T extends Partial<RouterInfo>>(info: T): T => {
  return {
    ...(info.params && { params: _sanitizeParams(info.params) }),
    ...(info.factoryAddr && { factoryAddr: info.factoryAddr.toLowerCase() }),
    ...(info.recipient && { recipient: info.recipient.toLowerCase() }),
    ...(info.feeAddr && { feeAddr: info.feeAddr.toLowerCase() }),
    ...(info.routerAddr && { routerAddr: info.routerAddr.toLowerCase() }),
  } as T;
};
