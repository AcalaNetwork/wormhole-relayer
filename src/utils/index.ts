export * from './configureEnv';
export * from './logger';
export * from './relay';
export * from './route';
export * from './utils';
export * from './validate';
export * from './wormhole';
export * from './address';
export * from './error';
