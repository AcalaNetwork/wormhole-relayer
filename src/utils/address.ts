import { evmToAddr32, nativeToAddr32 } from '@acala-network/asset-router/dist/utils';
import { isEvmAddress, isSubstrateAddress } from '@acala-network/eth-providers';

export const toAddr32 = (addr: string) => {
  if (!isEvmAddress(addr) || !isSubstrateAddress(addr)) {
    throw new Error(`address ${addr} is not a valid evm or substrate address`);
  }

  return isEvmAddress(addr)
    ? evmToAddr32(addr)
    : nativeToAddr32(addr);
};
