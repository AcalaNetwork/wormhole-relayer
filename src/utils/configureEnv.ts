import { ADDRESSES } from '@acala-network/asset-router/dist/consts';
import { CHAIN_ID_ACALA, CHAIN_ID_KARURA, ChainId } from '@certusone/wormhole-sdk';
import dotenv from 'dotenv';

import { ROUTER_CHAIN_ID } from './utils';

dotenv.config({ path: '.env' });

export type RelayerEnvironment = {
  supportedChains: ChainConfig[];
};

export type ChainConfig = {
  chainId: ROUTER_CHAIN_ID;
  ethRpc: string;
  walletPrivateKey: string;
  tokenBridgeAddr: string;
  feeAddr: string;
  factoryAddr: string;
  isTestnet: boolean;
};

const isTestnet = Boolean(Number(process.env.TESTNET_MODE ?? 1));

export function validateEnvironment(): RelayerEnvironment {
  const supportedChains: ChainConfig[] = [];
  supportedChains.push(configKarura());
  supportedChains.push(configAcala());

  return { supportedChains };
}

function configKarura(): ChainConfig {
  const requiredEnvVars = [
    'KARURA_ETH_RPC',
    'KARURA_PRIVATE_KEY',
  ];

  for (const envVar of requiredEnvVars) {
    if (!process.env[envVar]) {
      console.error(`Missing environment variable ${envVar}`);
      process.exit(1);
    }
  }

  const addresses = isTestnet
    ? ADDRESSES.KARURA_TESTNET
    : ADDRESSES.KARURA;

  return {
    chainId: CHAIN_ID_KARURA,
    ethRpc: process.env.KARURA_ETH_RPC!,
    walletPrivateKey: process.env.KARURA_PRIVATE_KEY!,
    isTestnet,
    ...addresses,
  };
}

function configAcala(): ChainConfig {
  const requiredEnvVars = [
    'ACALA_ETH_RPC',
    'ACALA_PRIVATE_KEY',
  ];

  for (const envVar of requiredEnvVars) {
    if (!process.env[envVar]) {
      console.error(`Missing environment variable ${envVar}`);
      process.exit(1);
    }
  }

  const addresses = isTestnet
    ? ADDRESSES.ACALA_TESTNET
    : ADDRESSES.ACALA;

  return {
    chainId: CHAIN_ID_ACALA,
    ethRpc: process.env.ACALA_ETH_RPC!,
    walletPrivateKey: process.env.ACALA_PRIVATE_KEY!,
    isTestnet,
    ...addresses,
  };
}

const env: RelayerEnvironment = validateEnvironment();

export const getChainConfig = (chainId: ChainId): ChainConfig | undefined => (
  env.supportedChains.find((x) => x.chainId === chainId)
);
