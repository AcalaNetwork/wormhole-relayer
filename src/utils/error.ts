import { serialize } from './utils';

export class NoRouteError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'NoRouteError';
  }
}
export class RelayerError extends Error {
  params: any;
  description: string;

  constructor(message: string, params?: any) {
    super(message);
    this.name = 'RelayerError';
    this.params = params;
    this.description = 'an error occurred!';
  }

  toJson() {
    return {
      msg: this.description,
      error: this.message,
      params: serialize(this.params),
    };
  }
}

export class RelayError extends RelayerError {
  constructor(message: string, params?: any) {
    super(message, params);
    this.name = 'RelayError';
    this.description = 'failed to relay this request!';
  }
};

export class RouteError extends RelayerError {
  constructor(message: string, params?: any) {
    super(message, params);
    this.name = 'RouteError';
    this.description = 'failed to route this request!';
  }
};
