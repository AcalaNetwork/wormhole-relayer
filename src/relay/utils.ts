import {
  hexToUint8Array,
  redeemOnEth,
  parseTransferPayload,
  importCoreWasm,
  ChainId,
  hexToNativeString,
} from '@certusone/wormhole-sdk';
import { ethers } from 'ethers';
import { ChainConfigInfo } from '../configureEnv';
import { EvmRpcProvider } from '@acala-network/eth-providers';
import {
  RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS,
} from './consts';
import axios from 'axios';
import { formatEther } from 'ethers/lib/utils';

interface VaaInfo {
  amount: bigint;
  originAddress: string;
  originChain: ChainId;
  targetAddress: string;
  targetChain: ChainId;
}
interface ShouldRelayResult {
  shouldRelay: boolean;
  msg: string;
} 

export const parseVaa = async (bytes: Uint8Array): Promise<VaaInfo> => {
  const { parse_vaa } = await importCoreWasm();
  const parsedVaa = parse_vaa(bytes);
  const buffered = Buffer.from(new Uint8Array(parsedVaa.payload));

  return parseTransferPayload(buffered);
};

export const shouldRelayVaa = (vaaInfo: VaaInfo): ShouldRelayResult => {
  const {
    amount,
    targetChain,
    originChain,
    originAddress,
  } = vaaInfo;
  const originAsset = hexToNativeString(originAddress, originChain);

  const res = shouldRelay({ targetChain, originAsset, amount });

  console.log('should relay: ', { targetChain, originAsset, amount, res });
  return res;
};

export const shouldRelay = ({
  targetChain,
  originAsset,
  amount: _amount,
}: {
  targetChain: number;
  originAsset: string;
  amount: bigint;
}): ShouldRelayResult => {
  const _noRelay = (msg: string): ShouldRelayResult => ({ shouldRelay: false, msg });

  if (!targetChain) return _noRelay('missing targetChain');
  if (!originAsset) return _noRelay('missing originAsset');
  if (!_amount) return _noRelay('missing transfer amount');

  let amount: bigint;
  try {
    amount = BigInt(_amount);
  } catch (e) {
    return _noRelay(`failed to parse amount: ${_amount}`);
  }

  const supported = RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS[targetChain];
  if (!supported) return _noRelay('target chain not supported');

  const minTransfer = supported[originAsset.toLowerCase()];
  if (!minTransfer) return _noRelay('token not supported');
  if (amount < BigInt(minTransfer)) return _noRelay(`transfer amount too small, expect at least ${minTransfer}`);

  return { shouldRelay: true, msg: '' };
};

export const relayEVM = async (
  chainConfigInfo: ChainConfigInfo,
  signedVAA: string,
  request: any,
  response: any
) => {
  const provider = EvmRpcProvider.from(chainConfigInfo.substrateNodeUrl);
  await provider.isReady();

  const signer = new ethers.Wallet(chainConfigInfo.walletPrivateKey, provider);

  const receipt = await redeemOnEth(
    chainConfigInfo.tokenBridgeAddress,
    signer,
    hexToUint8Array(signedVAA),
  );

  console.log('successfully redeemed on evm', receipt);
  response.status(200).json(receipt);
};

export const fetchBalance = async (address: string, url: string): Promise<number> => {
  try {
    const response = await axios.get(url, {
      data: {
        id: 0,
        jsonrpc: '2.0',
        method: 'eth_getBalance',
        params: [
          address,
          'latest',
        ],
      }
    });

    return Number(formatEther(response.data.result));
  } catch (e) {
    console.log('fetchBalance error: ', e);
    return -1;
  }
};
