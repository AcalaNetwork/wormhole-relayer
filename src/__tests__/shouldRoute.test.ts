import { CHAIN_ID_ETH } from '@certusone/wormhole-sdk';
import { describe, expect, it } from 'vitest';

import {
  GOERLI_USDC,
  PARA_ID,
  ROUTE_SUPPORTED_CHAINS_AND_ASSETS,
  ROUTE_SUPPORTED_CHAINS_AND_ASSETS_PROD,
} from '../consts';
import { Mainnet } from '../utils';
import {
  expectError,
  shouldRouteHoma,
  shouldRouteWormhole,
  shouldRouteXcm,
} from './testUtils';

describe.concurrent('/shouldRouteXcm', () => {
  const dest = '0x03010200a9200100d43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d';

  it('when should route (testnet)', async () => {
    for (const [destParaId, supportedTokens] of Object.entries(ROUTE_SUPPORTED_CHAINS_AND_ASSETS)) {
      for (const tokenAddr of supportedTokens) {
        let res = await shouldRouteXcm({
          dest,
          destParaId,
          originAddr: tokenAddr,
        });

        expect(res).to.deep.eq({
          data: {
            shouldRoute: true,
            routerAddr: '0x8341Cd8b7bd360461fe3ce01422fE3E24628262F',
            routerChainId: 11,
          },
        });

        res = await shouldRouteXcm({
          dest,
          destParaId,
          originAddr: tokenAddr.toUpperCase(),
        });

        expect(res).to.deep.eq({
          data: {
            shouldRoute: true,
            routerAddr: '0x8341Cd8b7bd360461fe3ce01422fE3E24628262F',
            routerChainId: 11,
          },
        });
      }
    }
  });

  it.skip('when should route (mainnet)', async () => {
    const resPendings = [] as Promise<any>[];
    for (const [destParaId, supportedTokens] of Object.entries(ROUTE_SUPPORTED_CHAINS_AND_ASSETS_PROD)) {
      for (const tokenAddr of supportedTokens) {
        let resPending = shouldRouteXcm({
          dest,
          destParaId,
          originAddr: tokenAddr,
        });

        resPendings.push(resPending);

        // should be case insensitive
        resPending = shouldRouteXcm({
          dest,
          destParaId,
          originAddr: tokenAddr.toUpperCase(),
        });

        resPendings.push(resPending);
      }
    }

    const allRes = await Promise.all(resPendings);
    expect(allRes.map(res => res.data)).toMatchSnapshot();
  });

  describe('when should not route', () => {
    it('when missing params', async () => {
      try {
        await shouldRouteXcm({
          destParaId: PARA_ID.BASILISK,
          originAddr: '0x07865c6e87b9f70255377e024ace6630c1eaa37f',
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['dest is a required field'], 400);
      }

      try {
        await shouldRouteXcm({
          dest,
          originAddr: '0x07865c6e87b9f70255377e024ace6630c1eaa37f',
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['destParaId is a required field'], 400);
      }

      try {
        await shouldRouteXcm({
          dest,
          destParaId: PARA_ID.BASILISK,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['originAddr is a required field'], 400);
      }
    });

    it('when bad params', async () => {
      const validArgs = {
        dest,
        destParaId: PARA_ID.BASILISK,
        originAddr: '0x07865c6e87b9f70255377e024ace6630c1eaa37f',
      };

      let res = await shouldRouteXcm({
        ...validArgs,
        destParaId: 1111,
      });
      expect(res).to.deep.eq({
        data: {
          shouldRoute: false,
          msg: 'unsupported dest parachain: 1111',
        },
      });

      const unsupportedToken = '0x07865c6e87b9f70255377e024ace6630c1e00000';
      res = await shouldRouteXcm({
        ...validArgs,
        originAddr: unsupportedToken,
      });
      expect(res).to.deep.eq({
        data: {
          shouldRoute: false,
          msg: `unsupported token on dest parachin 2090. Token origin address: ${unsupportedToken}`,
        },
      });

      // TODO: need to validate dest?
      // res = await shouldRouteXcm({
      //   ...validArgs,
      //   dest: '0xabcd'
      // });
      // expect(res.data.shouldRoute).to.equal(false);
      // expect(res.data.msg).to.contain('unsupported router chainId: 8');
    });
  });
});

describe.concurrent('/shouldRouteWormhole', () => {
  it('when should route (testnet)', async () => {
    for (const supportedTokens of Object.values(ROUTE_SUPPORTED_CHAINS_AND_ASSETS)) {
      for (const tokenAddr of supportedTokens) {
        let res = await shouldRouteWormhole({
          originAddr: tokenAddr,
          targetChainId: String(CHAIN_ID_ETH),
          destAddr: '0x0085560b24769dAC4ed057F1B2ae40746AA9aAb6',
          fromParaId: PARA_ID.BASILISK,
        });

        expect(res).to.deep.eq({
          data: {
            shouldRoute: true,
            routerAddr: '0xC8a0596848966f61be4cd1875373d2728e162eE2',
          },
        });

        // should be case insensitive
        res = await shouldRouteWormhole({
          originAddr: tokenAddr,
          targetChainId: String(CHAIN_ID_ETH),
          destAddr: '0x0085560b24769dAC4ed057F1B2ae40746AA9aAb6',
          fromParaId: PARA_ID.BASILISK,
        });

        expect(res).to.deep.eq({
          data: {
            shouldRoute: true,
            routerAddr: '0xC8a0596848966f61be4cd1875373d2728e162eE2',
          },
        });
      }
    }
  });

  it.skip('when should route (mainnet)', async () => {
    const resPendings = [] as Promise<any>[];
    for (const supportedTokens of Object.values(ROUTE_SUPPORTED_CHAINS_AND_ASSETS_PROD)) {
      for (const tokenAddr of supportedTokens) {
        let resPending = shouldRouteWormhole({
          originAddr: tokenAddr,
          targetChainId: String(CHAIN_ID_ETH),
          destAddr: '0x0085560b24769dAC4ed057F1B2ae40746AA9aAb6',
          fromParaId: PARA_ID.BASILISK,
        });

        resPendings.push(resPending);

        // should be case insensitive
        resPending = shouldRouteWormhole({
          originAddr: tokenAddr,
          targetChainId: String(CHAIN_ID_ETH),
          destAddr: '0x0085560b24769dAC4ed057F1B2ae40746AA9aAb6',
          fromParaId: PARA_ID.BASILISK,
        });

        resPendings.push(resPending);
      }
    }

    const allRes = await Promise.all(resPendings);
    for (const res of allRes) {
      // router addr should only change if fee addr changed
      expect(res).to.deep.eq({
        data: {
          shouldRoute: true,
          routerAddr: '0x5E0fE43f1eeaca029390D061DE2e6377B5eaEE3a',
        },
      });
    }
  });

  describe('when should not route', () => {
    it('when missing params', async () => {
      try {
        await shouldRouteWormhole({
          targetChainId: String(CHAIN_ID_ETH),
          destAddr: '0x0085560b24769dAC4ed057F1B2ae40746AA9aAb6',
          fromParaId: PARA_ID.BASILISK,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['originAddr is a required field'], 400);
      }

      try {
        await shouldRouteWormhole({
          originAddr: GOERLI_USDC,
          destAddr: '0x0085560b24769dAC4ed057F1B2ae40746AA9aAb6',
          fromParaId: PARA_ID.BASILISK,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['targetChainId is a required field'], 400);
      }

      try {
        await shouldRouteWormhole({
          originAddr: GOERLI_USDC,
          targetChainId: String(CHAIN_ID_ETH),
          fromParaId: PARA_ID.BASILISK,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['destAddr is a required field'], 400);
      }

      try {
        await shouldRouteWormhole({
          originAddr: GOERLI_USDC,
          targetChainId: String(CHAIN_ID_ETH),
          destAddr: '0x0085560b24769dAC4ed057F1B2ae40746AA9aAb6',
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['fromParaId is a required field'], 400);
      }
    });

    it('when bad params', async () => {
      const validArgs = {
        originAddr: GOERLI_USDC,
        targetChainId: String(CHAIN_ID_ETH),
        destAddr: '0x0085560b24769dAC4ed057F1B2ae40746AA9aAb6',
        fromParaId: PARA_ID.BASILISK,
      };

      let res = await shouldRouteWormhole({
        ...validArgs,
        fromParaId: 1111,
      });
      expect(res).to.deep.eq({
        data: {
          shouldRoute: false,
          msg: 'unsupported origin parachain: 1111',
        },
      });

      const unsupportedToken = '0x07865c6e87b9f70255377e024ace6630c1e00000';
      res = await shouldRouteWormhole({
        ...validArgs,
        originAddr: '0x07865c6e87b9f70255377e024ace6630c1e00000',
      });
      expect(res).to.deep.eq({
        data: {
          shouldRoute: false,
          msg: `origin token ${unsupportedToken} not supported on router chain 11`,
        },
      });
    });
  });
});

describe.concurrent.skip('/shouldRouteHoma', () => {
  const destAddr = '0x75E480dB528101a381Ce68544611C169Ad7EB342';
  const destAddrSubstrate = '23AdbsfRysaabyrWS2doCFsKisvt7dGbS3wQFXRS6pNbQY8G';

  describe('when should route', async () => {
    it('to evm address', async () => {
      // for (const network of [Object.values(Mainnet)]) {
      for (const chain of ['acala']) {
        let res = await shouldRouteHoma({
          destAddr,
          chain,
        });

        expect(res).toMatchInlineSnapshot(`
        {
          "data": {
            "routerAddr": "0xa013818BBddc5d2d55ab9cCD50759b3B1953d6cd",
            "shouldRoute": true,
          },
        }
      `);

        // should be case insensitive
        res = await shouldRouteHoma({
          destAddr: destAddr.toLocaleLowerCase(),
          chain,
        });

        expect(res).toMatchInlineSnapshot(`
        {
          "data": {
            "routerAddr": "0xa013818BBddc5d2d55ab9cCD50759b3B1953d6cd",
            "shouldRoute": true,
          },
        }
      `);
      }
    });

    it('to substrate address', async () => {
    // for (const network of [Object.values(Mainnet)]) {
      for (const chain of ['acala']) {
        let res = await shouldRouteHoma({
          destAddr: destAddrSubstrate,
          chain,
        });

        expect(res).toMatchInlineSnapshot(`
          {
            "data": {
              "routerAddr": "0xfD6143c380706912a04230f22cF92c402561820e",
              "shouldRoute": true,
            },
          }
        `);

        // should be case insensitive
        res = await shouldRouteHoma({
          destAddr: destAddrSubstrate.toLocaleLowerCase(),
          chain,
        });

        expect(res).toMatchInlineSnapshot(`
          {
            "data": {
              "msg": "address 23adbsfrysaabyrws2docfskisvt7dgbs3wqfxrs6pnbqy8g is not a valid evm or substrate address",
              "shouldRoute": false,
            },
          }
        `);
      }
    });
  });

  describe('when should not route', () => {
    it('when missing params', async () => {
      try {
        await shouldRouteHoma({
          destAddr,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['chain is a required field'], 400);
      }

      try {
        await shouldRouteHoma({
          chain: Mainnet.Acala,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['destAddr is a required field'], 400);
      }

      try {
        await shouldRouteHoma({
          chain: 'mandala',
          destAddr,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['chain must be one of the following values: acala, karura'], 400);
      }
    });

    it('when bad params', async () => {
      const res = await shouldRouteHoma({
        chain: Mainnet.Acala,
        destAddr: '0xaaaaaaaaaa',
      });
      expect(res).toMatchInlineSnapshot(`
        {
          "data": {
            "msg": "address 0xaaaaaaaaaa is not a valid evm or substrate address",
            "shouldRoute": false,
          },
        }
      `);
    });
  });
});
