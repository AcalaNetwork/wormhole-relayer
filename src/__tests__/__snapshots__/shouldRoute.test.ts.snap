// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/shouldRouteXcm > when should route (mainnet) 1`] = `
[
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0xE0b40e0CeC44DAD0E1Dbf7968DF4CbdA67b9dDA4",
      "routerChainId": 11,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
  {
    "data": {
      "routerAddr": "0x8D6F87eDdC3F3d393659A47351a6Bb9AdB48b244",
      "routerChainId": 12,
      "shouldRoute": true,
    },
  },
]
`;
